### PR TITLE
/addUnknowns endpoint functionality.

### PR DESCRIPTION
**Pull request by Ron-Madan**
Here is the state of the courts before the endpoint:

1-Old

2-Old

3-Old

4-Old

You can see that courts 1 & 3 are empty, and courts 2 & 4 are occupied.

Now, we call the endpoint. We inform the backend that courts 1,2 and 3 are occupied. This means courts 1 and 3 should now have unknown names. Court 2 should stay the same. Lastly, court 4 should reset to empty.

Postman

We can see the changes reflected in Firestore:
1-New

2-New

3-New

4-New

Note how the timestamps for 1 and 3 have changed to time of the endpoint request with the random adjustment. Also, note how the difference in their timestamps is 20 seconds, which could have arisen from a -40, -20 / -30, -10 / -20, 0 adjustment selection.

### CODE REVIEW

File: backend/routes/addUnknowns.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    if (!location || !Array.isArray(occupiedCourts)) {
+      return res.status(400).json({ message: 'Location and occupied court information is required.' });
+    }
+
+    // Retrieve active court information for the given location
+    const activePlayersSnapshot = await admin.firestore()
+      .collection('locations')
+      .doc(location)
+      .collection('activePlayers')
+      .orderBy('courtNumber')
```
Comment (LeandroHamaguchi) [inline backend/routes/addUnknowns.js:19]: Does it need to be ordered in this case? I believe that removing .orderBy() would improve the runtime.

File: backend/routes/addUnknowns.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+      const adjustedTime = new Date(now);
+      adjustedTime.setSeconds(now.getSeconds() + seconds);
+      return adjustedTime;
+    });
+    shuffleArray(timestamps);
+    let lastTimestampIdx = 0;
+
+    // Update each document based on user information
+    const batch = admin.firestore().batch();
+    activePlayersSnapshot.docs.forEach(doc => {
```
Comment (LeandroHamaguchi) [inline backend/routes/addUnknowns.js:40]: I saw in this [documentation](https://firebase.google.com/docs/reference/node/firebase.firestore.QuerySnapshot#foreach) that there is a forEach method for the snapshot. I think there's no need to have `.docs`.

File: backend/routes/addUnknowns.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+      .get();
+
+    // Check for empty active player snapshot (as always filled with Empty/Unknown/Player objects)
+    if (activePlayersSnapshot.empty) {
+      return res.status(404).json({ message: 'Invalid location.' });
+    }
+
+    // Creating shuffled timestamp adjustments
+    const now = new Date();
+    const adjustments = [-40, -30, -20, -10, 0];
```
[the remark this answers is not in the document]
Comment (Ron-Madan) [reply backend/routes/addUnknowns.js]: Hey Leandro, thanks for the feedback!

"(occupiedCourts.length - i)" may cause an issue if the user only selects court 3 as the length would be 1. However, your approach makes sense so I have used each index of the list to create the adjustments:

"occupiedCourts.map((_, index) => -10 * (index + 1));" So we'd get -10, -20, ... -10n

File: backend/routes/addUnknowns.js [hunk 4]
```diff
@@ -37,41 +36,37 @@ router.post('/addUnknowns', async (req, res) => {
...
-          const updateObj = {
-            playerWaiting: false,
-            firebaseUID: newName,
-            nickname: newName,
-          };
+        const updateObj = {
+          playerWaiting: false,
+          firebaseUID: newName,
+          nickname: newName,
+        };
```
Comment (LeandroHamaguchi) [inline backend/routes/addUnknowns.js:67]: Refactoring this if statement is a great addition! However, I believe there are 2 other possibilities when `!isPhysicallyOccupied && isVirtuallyEmpty` (No one on the court both physically and virtually) and `isPhysicallyOccupied && !isVirtuallyEmpty` (Someone is on the court both physically and virtually). In both cases, it wouldn't make sense to call this endpoint, but it's still possible to (if there are no restrictions in the frontend). Should these cases also be handled and return an error?

Comment (Ron-Madan) [reply backend/routes/addUnknowns.js:67]: Hey Leandro, good observation. Regarding these 2 cases, this means that our database has the up to date information on those courts. Our main focus is the other 2 cases currently identified as that means there is a mismatch.

**Review by LeandroHamaguchi — APPROVED** [review]
LGTM

**Pull request merged**